
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>geppy.tools package &#8212; geppy 0.1.2 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/pydoctheme.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
 
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-15144370-3']);
  _gaq.push(['_trackPageview']);
</script>
<script type="text/javascript" src="_static/copybutton.js"></script>

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
    <li><a href="https://github.com/ShuhuaGao/geppy"><img src="_static/geppy-icon.png" style="height: 30px; vertical-align: middle; padding-right: 1em;" />Project Homepage</a> &#187;</li>
	<li><a href="index.html">geppy 0.1.2 documentation</a> &#187; </li>

        <li class="nav-item nav-item-this"><a href="">geppy.tools package</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="geppy-tools-package">
<h1>geppy.tools package<a class="headerlink" href="#geppy-tools-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-geppy.tools.crossover">
<span id="geppy-tools-crossover-module"></span><h2>geppy.tools.crossover module<a class="headerlink" href="#module-geppy.tools.crossover" title="Permalink to this headline">¶</a></h2>
<p>The module <code class="xref py py-mod docutils literal notranslate"><span class="pre">crossover</span></code> provides crossover (mating, recombination) related genetic modifications in GEP,
including one-point and two-point crossover, and gene crossover between multigenic chromosomes.
Please refer to Chapter 3 of <a class="reference internal" href="#fc2006" id="id1"><span>[FC2006]</span></a> for more details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All the recombination operators including <a class="reference internal" href="#geppy.tools.crossover.crossover_one_point" title="geppy.tools.crossover.crossover_one_point"><code class="xref py py-func docutils literal notranslate"><span class="pre">crossover_one_point()</span></code></a>, <a class="reference internal" href="#geppy.tools.crossover.crossover_two_point" title="geppy.tools.crossover.crossover_two_point"><code class="xref py py-func docutils literal notranslate"><span class="pre">crossover_two_point()</span></code></a>, and
<a class="reference internal" href="#geppy.tools.crossover.crossover_gene" title="geppy.tools.crossover.crossover_gene"><code class="xref py py-func docutils literal notranslate"><span class="pre">crossover_gene()</span></code></a> can be applied to both <a class="reference internal" href="geppy.core.html#geppy.core.entity.Gene" title="geppy.core.entity.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a>
and <a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>.</p>
</div>
<dl class="citation">
<dt class="label" id="fc2006"><span class="brackets">FC2006</span><span class="fn-backref">(<a href="#id1">1</a>,<a href="#id2">2</a>,<a href="#id3">3</a>,<a href="#id4">4</a>)</span></dt>
<dd><p>Ferreira, Cândida. Gene expression programming: mathematical modeling by an artificial
intelligence. Vol. 21. Springer, 2006.</p>
</dd>
</dl>
<dl class="py function">
<dt id="geppy.tools.crossover.crossover_gene">
<code class="sig-prename descclassname">geppy.tools.crossover.</code><code class="sig-name descname">crossover_gene</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ind1</span></em>, <em class="sig-param"><span class="n">ind2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/crossover.html#crossover_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.crossover.crossover_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Entire genes are exchanged between two parent chromosomes. The two individuals are modified in place, and the two
children are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ind1</strong> – The first individual (chromosome) participating in the crossover.</p></li>
<li><p><strong>ind2</strong> – The second individual (chromosome) participating in the crossover.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of two children individuals</p>
</dd>
</dl>
<p>This operation has no effect if the chromosome has only one gene. Typically, a gene recombination rate
around 0.2 is used.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.crossover.crossover_one_point">
<code class="sig-prename descclassname">geppy.tools.crossover.</code><code class="sig-name descname">crossover_one_point</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ind1</span></em>, <em class="sig-param"><span class="n">ind2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/crossover.html#crossover_one_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.crossover.crossover_one_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute one-point recombination of two individuals. The two individuals are modified in place, and the two children
are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ind1</strong> – The first individual (chromosome) participating in the crossover.</p></li>
<li><p><strong>ind2</strong> – The second individual (chromosome) participating in the crossover.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of two children individuals.</p>
</dd>
</dl>
<p>Note the crossover can happen at any point across the whole chromosome and thus entire genes may be also exchanged
between the two parents if they are multigenic chromosomes.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.crossover.crossover_two_point">
<code class="sig-prename descclassname">geppy.tools.crossover.</code><code class="sig-name descname">crossover_two_point</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ind1</span></em>, <em class="sig-param"><span class="n">ind2</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/crossover.html#crossover_two_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.crossover.crossover_two_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute two-point recombination of two individuals. The two individuals are modified in place, and the two children
are returned. The materials between two randomly chosen points are swapped to generate two children.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ind1</strong> – The first individual (chromosome) participating in the crossover.</p></li>
<li><p><strong>ind2</strong> – The second individual (chromosome) participating in the crossover.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of two individuals.</p>
</dd>
</dl>
<p>Note the crossover can happen at any point across the whole chromosome and thus entire genes may be also exchanged
between the two parents if they are multigenic chromosomes.</p>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.generator">
<span id="geppy-tools-generator-module"></span><h2>geppy.tools.generator module<a class="headerlink" href="#module-geppy.tools.generator" title="Permalink to this headline">¶</a></h2>
<p>This module <code class="xref py py-mod docutils literal notranslate"><span class="pre">generator</span></code> provides functionality to generate a genome for a gene. That is, choose functions and
terminals randomly from a given primitive set to form a linear form gene expression.</p>
<dl class="py function">
<dt id="geppy.tools.generator.generate_dc">
<code class="sig-prename descclassname">geppy.tools.generator.</code><code class="sig-name descname">generate_dc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rnc_array_length</span></em>, <em class="sig-param"><span class="n">dc_length</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/generator.html#generate_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.generator.generate_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a Dc domain for a RNC array of size <em>rnc_array_length</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rnc_array_length</strong> – length of the RNC array</p></li>
<li><p><strong>dc_length</strong> – length of the Dc domain</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of integers of length <em>dc_length</em>, each element in range <cite>[0, rnc_array_length - 1]</cite></p>
</dd>
</dl>
<p>Refer to Chapter 5 of <a class="reference internal" href="#fc2006" id="id2"><span>[FC2006]</span></a> for more details.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.generator.generate_genome">
<code class="sig-prename descclassname">geppy.tools.generator.</code><code class="sig-name descname">generate_genome</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pset</span></em>, <em class="sig-param"><span class="n">head_length</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/generator.html#generate_genome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.generator.generate_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a genome with the given primitive set <em>pset</em> and the specified head domain length <em>head_length</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pset</strong> – a primitive set</p></li>
<li><p><strong>head_length</strong> – length of the head domain</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of symbols representing a genome</p>
</dd>
</dl>
<p>Supposing the maximum arity of functions in <em>pset</em> is <em>max_arity</em>, then the tail length is automatically
determined to be <code class="docutils literal notranslate"><span class="pre">tail_length</span> <span class="pre">=</span> <span class="pre">head_length</span> <span class="pre">*</span> <span class="pre">(max_arity</span> <span class="pre">-</span> <span class="pre">1)</span> <span class="pre">+</span> <span class="pre">1</span></code>.</p>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.mutation">
<span id="geppy-tools-mutation-module"></span><h2>geppy.tools.mutation module<a class="headerlink" href="#module-geppy.tools.mutation" title="Permalink to this headline">¶</a></h2>
<p>The module <code class="xref py py-mod docutils literal notranslate"><span class="pre">mutation</span></code> provides mutation related genetic modification operators in GEP, including point-mutation,
transposition and inversion. Please refer to Chapter 3 of <a class="reference internal" href="#fc2006" id="id3"><span>[FC2006]</span></a> for more details.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Operators that can be applied to both <a class="reference internal" href="geppy.core.html#geppy.core.entity.Gene" title="geppy.core.entity.Gene"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gene</span></code></a> and <a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>:
<a class="reference internal" href="#geppy.tools.mutation.mutate_uniform" title="geppy.tools.mutation.mutate_uniform"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate_uniform()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.invert" title="geppy.tools.mutation.invert"><code class="xref py py-func docutils literal notranslate"><span class="pre">invert()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.is_transpose" title="geppy.tools.mutation.is_transpose"><code class="xref py py-func docutils literal notranslate"><span class="pre">is_transpose()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.ris_transpose" title="geppy.tools.mutation.ris_transpose"><code class="xref py py-func docutils literal notranslate"><span class="pre">ris_transpose()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.gene_transpose" title="geppy.tools.mutation.gene_transpose"><code class="xref py py-func docutils literal notranslate"><span class="pre">gene_transpose()</span></code></a></p></li>
<li><p>Operators that end with ‘_dc’ are specially designed to handle the Dc domain in
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>.</p></li>
</ol>
</div>
<dl class="py function">
<dt id="geppy.tools.mutation.gene_transpose">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">gene_transpose</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#gene_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.gene_transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform gene transposition in place</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>An entire gene is selected randomly and exchanged with the first gene in the chromosome.
Obviously, this operation only makes sense if the chromosome is a multigenic one. Typically, the gene transposition
rate is set around 0.1.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.invert">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">invert</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#invert"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.invert" title="Permalink to this definition">¶</a></dt>
<dd><p>A gene is randomly chosen, and afterwards a subsequence within this gene’s head domain is randomly selected
and inverted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>Typically, a small inversion rate of 0.1 is used.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.invert_dc">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">invert_dc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#invert_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.invert_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Dc-specific inversion. A gene is randomly chosen, and afterwards a subsequence within this gene’s Dc domain is
randomly selected and inverted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>Typically, a small inversion rate of 0.1 is used.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.is_transpose">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">is_transpose</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#is_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.is_transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform IS transposition in place</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>An Insertion Sequence (IS) is a random chosen segment across the chromosome, and then the
IS is copied to be inserted at another position in the head of gene, except the start position. Note that an IS from
one gene may be copied to a different gene.
Typically, the IS transposition rate is set around 0.1.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.mutate_rnc_array_dc">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">mutate_rnc_array_dc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em>, <em class="sig-param"><span class="n">rnc_gen</span></em>, <em class="sig-param"><span class="n">ind_pb</span><span class="o">=</span><span class="default_value">'1p'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_rnc_array_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_rnc_array_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct mutation of RNCs, which changes the values in a gene’s RNC array
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc.rnc_array" title="geppy.core.entity.GeneDc.rnc_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rnc_array()</span></code></a> randomly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a>.</p></li>
<li><p><strong>rnc_gen</strong> – callable, which returns a random numerical constant by calling <code class="docutils literal notranslate"><span class="pre">rnc_gen()</span></code>.</p></li>
<li><p><strong>ind_pb</strong> – float or str, default ‘1p’. Use a float number to specify the probability of each RNC in the array
being mutated. Alternatively, if a str ending with ‘p’ is given in the form like ‘xp’,  where
‘x’ is a numerical number, then it specifies how many point mutations are expected for the RNC arrays
in <em>individual</em>. For example, if there are <em>d</em> RNCs inside the arrays in total in <em>individual</em>,
then ‘1.5p’ is approximately equal to <cite>ind_pb = 1.5 / d</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>The genetic operators <a class="reference internal" href="#geppy.tools.mutation.mutate_uniform_dc" title="geppy.tools.mutation.mutate_uniform_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate_uniform_dc()</span></code></a>, <a class="reference internal" href="#geppy.tools.mutation.transpose_dc" title="geppy.tools.mutation.transpose_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">transpose_dc()</span></code></a> and <a class="reference internal" href="#geppy.tools.mutation.invert_dc" title="geppy.tools.mutation.invert_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">invert_dc()</span></code></a> actually only move the
random numerical constants around without generating new numerical values. This method <a class="reference internal" href="#geppy.tools.mutation.mutate_rnc_array_dc" title="geppy.tools.mutation.mutate_rnc_array_dc"><code class="xref py py-func docutils literal notranslate"><span class="pre">mutate_rnc_array_dc()</span></code></a>
can replace the value of a particular numerical constant by another in the
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc.rnc_array" title="geppy.core.entity.GeneDc.rnc_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rnc_array()</span></code></a>.</p>
<p>Refer to section 5.4.4 of <a class="reference internal" href="#fc2006" id="id4"><span>[FC2006]</span></a> for more details.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.mutate_uniform">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">mutate_uniform</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em>, <em class="sig-param"><span class="n">pset</span></em>, <em class="sig-param"><span class="n">ind_pb</span><span class="o">=</span><span class="default_value">'2p'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_uniform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_uniform" title="Permalink to this definition">¶</a></dt>
<dd><p>Uniform point mutation. For each symbol (primitive) in <em>individual</em>, change it to another randomly chosen symbol
from <em>pset</em> with the probability <em>indpb</em>. A symbol may be a function or a terminal.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, the chromosome to be mutated.</p></li>
<li><p><strong>pset</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">PrimitiveSet</span></code>, a primitive set</p></li>
<li><p><strong>ind_pb</strong> – float or str, default ‘2p’. Probability of mutating each symbol.
If <em>ind_pb</em> is given as a string ending with ‘p’,
then it indicates the expected number of point mutations among all the symbols in <em>individual</em>. For example,
if the total length of each gene of <em>individual</em> is <cite>l</cite> and there are <cite>m</cite> genes in total, then by passing
<cite>ind_pb=’1.5p’</cite> we specify approximately <cite>ind_pb=1.5/(l*m)</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of one chromosome</p>
</dd>
</dl>
<p>It is typical to set a mutation rate <em>indpb</em> equivalent to two one-point mutations per chromosome. That is,
<code class="docutils literal notranslate"><span class="pre">indpb</span> <span class="pre">=</span> <span class="pre">2</span> <span class="pre">/</span> <span class="pre">len(chromosome)</span> <span class="pre">*</span> <span class="pre">len(gene)</span></code>.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.mutate_uniform_dc">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">mutate_uniform_dc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em>, <em class="sig-param"><span class="n">ind_pb</span><span class="o">=</span><span class="default_value">'1p'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_uniform_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_uniform_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>Dc-specific mutation. This operator changes one index stored in the Dc domain to another index in place. The indices
in the Dc domain can later be used to retrieve numerical constants from the gene associated
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc.rnc_array" title="geppy.core.entity.GeneDc.rnc_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rnc_array()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a></p></li>
<li><p><strong>ind_pb</strong> – float or str, default ‘1p’. Probability of mutating each index/position in the Dc domain. If given
in str like ‘xp’, then <cite>x</cite> point mutations for expected for the Dc domains in total of <em>individual</em>. For
instance, if there are <cite>d</cite> Dc elements in the whole chromosome, then <cite>1.5p</cite> is equal to <cite>ind_pb = 1.5 / d</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one chromosome</p>
</dd>
</dl>
<p>It is typical to set a mutation rate <em>indpb</em> equivalent to two one-point mutations per chromosome. That is,
<code class="docutils literal notranslate"><span class="pre">indpb</span> <span class="pre">=</span> <span class="pre">2</span> <span class="pre">/</span> <span class="pre">len(chromosome)</span> <span class="pre">*</span> <span class="pre">len(gene)</span></code>.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.mutate_uniform_ephemeral">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">mutate_uniform_ephemeral</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em>, <em class="sig-param"><span class="n">ind_pb</span><span class="o">=</span><span class="default_value">'1p'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#mutate_uniform_ephemeral"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.mutate_uniform_ephemeral" title="Permalink to this definition">¶</a></dt>
<dd><p>This operator is specially designed for ephemeral numerical constant (ENC) mutation.
It attempts to change the value of the ephemeral constants in the <em>individual</em> if any by calling <cite>e.update_value()</cite>,
where <cite>e</cite> is an ephemeral inside <em>individual</em>. See <a class="reference internal" href="geppy.core.html#geppy.core.symbol.EphemeralTerminal.update_value" title="geppy.core.symbol.EphemeralTerminal.update_value"><code class="xref py py-meth docutils literal notranslate"><span class="pre">update_value()</span></code></a> for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</p></li>
<li><p><strong>ind_pb</strong> – float or str, default ‘1p’. If a float is given, then <em>ind_pb</em> specifies the probability of each
ephemeral in the <em>individual</em> being mutated. If a str ending with ‘p’ is given in the form like ‘xp’,  where
‘x’ is a numerical number, then it
specifies how many point mutations are expected for the ephemeral constants. For example, if there are <em>d</em>
ephemeral constants present in total in <em>individual</em>, then ‘1.5p’ is approximately equal to <cite>ind_pb = 1.5 / d</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.ris_transpose">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">ris_transpose</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#ris_transpose"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.ris_transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform RIS transposition in place</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>A Root Insertion Sequence (RIS) is a segment of consecutive elements that starts with a
function. Once an RIS is randomly chosen, it is copied and inserted at the root (first position) of a gene. Note
that an RIS from one gene may be copied to a different gene.
Typically, the RIS transposition rate is set around 0.1.</p>
</dd></dl>

<dl class="py function">
<dt id="geppy.tools.mutation.transpose_dc">
<code class="sig-prename descclassname">geppy.tools.mutation.</code><code class="sig-name descname">transpose_dc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/mutation.html#transpose_dc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.mutation.transpose_dc" title="Permalink to this definition">¶</a></dt>
<dd><p>This operator randomly chooses the chromosome, the gene with its respective Dc to be subjected to transposition,
the start and termination points of the transposon, and the target site; then it inserts a copy of the transposon
from the place of origin to the target site, while the old content at the target site are shifted to the right
and the length of the Dc domain is maintained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> – <a class="reference internal" href="geppy.core.html#geppy.core.entity.Chromosome" title="geppy.core.entity.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>, a chromosome, which contains genes of type
<a class="reference internal" href="geppy.core.html#geppy.core.entity.GeneDc" title="geppy.core.entity.GeneDc"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneDc</span></code></a></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of one individual</p>
</dd>
</dl>
<p>Typically, a small transposition rate of 0.1 is used. Note that a transposon from one gene’s Dc may be copied and
inserted into another gene’s Dc, which means the genes in a multigenic chromosome <em>individual</em>
must have an identical structure.</p>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.parser">
<span id="geppy-tools-parser-module"></span><h2>geppy.tools.parser module<a class="headerlink" href="#module-geppy.tools.parser" title="Permalink to this headline">¶</a></h2>
<p>This module <a class="reference external" href="https://docs.python.org/3/library/parser.html#module-parser" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">parser</span></code></a> provides functionality for compiling an individual (a chromosome) in GEP into an executable
lambda function in Python for subsequent fitness evaluation.</p>
<div class="admonition-todo admonition" id="id5">
<p class="admonition-title">Todo</p>
<p>Parse an individual into codes of other languages, such as C++/Java, for deployment in an industrial environment.</p>
</div>
<dl class="py function">
<dt id="geppy.tools.parser.compile_">
<code class="sig-prename descclassname">geppy.tools.parser.</code><code class="sig-name descname">compile_</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span></em>, <em class="sig-param"><span class="n">pset</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/geppy/tools/parser.html#compile_"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#geppy.tools.parser.compile_" title="Permalink to this definition">¶</a></dt>
<dd><p>Compile the individual into a Python lambda expression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individual</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code>, a chromosome</p></li>
<li><p><strong>pset</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">PrimitiveSet</span></code>, a primitive set</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a function if the primitive set <em>pset</em> has any inputs (arguments), which can later be called with
specific parameter values; otherwise, a numerical result obtained from evaluation.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-geppy.tools.toolbox">
<span id="geppy-tools-toolbox-module"></span><h2>geppy.tools.toolbox module<a class="headerlink" href="#module-geppy.tools.toolbox" title="Permalink to this headline">¶</a></h2>
<p>The module <code class="xref py py-mod docutils literal notranslate"><span class="pre">toolbox</span></code> mainly provides a class <code class="xref py py-class docutils literal notranslate"><span class="pre">Toolbox</span></code> on the basis of <a class="reference external" href="https://deap.readthedocs.io/en/master/api/base.html#deap.base.Toolbox" title="(in DEAP v1.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">deap.base.Toolbox</span></code></a> for GEP
operator registration, which is then extensively used in the builtin GEP algorithms. For user-defined custom algorithms,
they should also accept a toolbox instance and retrieve the operators and their associated probability from the toolbox
instance.</p>
</div>
<div class="section" id="module-geppy.tools">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-geppy.tools" title="Permalink to this headline">¶</a></h2>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">geppy.tools package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-geppy.tools.crossover">geppy.tools.crossover module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.generator">geppy.tools.generator module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.mutation">geppy.tools.mutation module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.parser">geppy.tools.parser module</a></li>
<li><a class="reference internal" href="#module-geppy.tools.toolbox">geppy.tools.toolbox module</a></li>
<li><a class="reference internal" href="#module-geppy.tools">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/geppy.tools.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
    <li><a href="https://github.com/ShuhuaGao/geppy"><img src="_static/geppy-icon.png" style="height: 30px; vertical-align: middle; padding-right: 1em;" />Project Homepage</a> &#187;</li>
	<li><a href="index.html">geppy 0.1.2 documentation</a> &#187; </li>

        <li class="nav-item nav-item-this"><a href="">geppy.tools package</a></li> 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2018-2020, Shuhua Gao.
    <br />
    Built on None.
    <a href="bugs.html">Found a bug</a>?
    <br />
    Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 3.3.1.
</div>
<script type="text/javascript">
  (function() {
     var ga = document.createElement('script');
         ga.src = ('https:' == document.location.protocol ?
                'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    ga.setAttribute('async', 'true');
    document.documentElement.firstChild.appendChild(ga);
  })();
</script>

  </body>
</html>